File: src/lib/schedule/doses.ts
import { addHours, addMinutes, isBefore, isEqual } from "date-fns";
import type { ParsedTreatment } from "@/lib/types";

export type PlannedDose = {
  scheduledAt: Date;
  remindAt: Date;
};

function parseAnchorTimes(anchorTimes: string[], firstDoseAt: Date): Date[] {
  if (anchorTimes.length === 0) return [firstDoseAt];

  const year = firstDoseAt.getFullYear();
  const month = firstDoseAt.getMonth();
  const day = firstDoseAt.getDate();

  const anchors = anchorTimes.map((hhmm) => {
    const [h, m] = hhmm.split(":").map(Number);
    return new Date(year, month, day, h, m, 0, 0);
  });

  anchors.sort((a, b) => a.getTime() - b.getTime());
  return anchors;
}

export function estimateDoseCount(parsed: ParsedTreatment): number {
  return buildDosePlan(parsed).length;
}

export function buildDosePlan(parsed: ParsedTreatment): PlannedDose[] {
  const firstDoseAt = new Date(parsed.firstDoseAt);
  const endsAt = new Date(parsed.endsAt);
  const remindBefore = parsed.remindBeforeMinutes ?? 15;

  if (!(endsAt > firstDoseAt)) {
    throw new Error("A data de fim precisa ser depois da primeira dose.");
  }

  const doses: PlannedDose[] = [];
  const pushDose = (scheduledAt: Date) => {
    if (scheduledAt > endsAt) return;
    if (doses.some((d) => isEqual(d.scheduledAt, scheduledAt))) return;
    doses.push({
      scheduledAt,
      remindAt: addMinutes(scheduledAt, -remindBefore),
    });
  };

  if (parsed.frequencyEveryHours) {
    let cursor = firstDoseAt;
    while (cursor <= endsAt) {
      pushDose(cursor);
      cursor = addHours(cursor, parsed.frequencyEveryHours);
      if (doses.length > 500) break;
    }
    return doses.sort((a, b) => a.scheduledAt.getTime() - b.scheduledAt.getTime());
  }

  const anchors = parseAnchorTimes(parsed.anchorTimes, firstDoseAt);
  const timesPerDay = parsed.timesPerDay ?? anchors.length;
  const dailyTimes = anchors.slice(0, Math.max(timesPerDay, 1));

  let dayOffset = 0;
  while (true) {
    let addedInDay = 0;
    for (const base of dailyTimes) {
      const scheduledAt = addHours(base, dayOffset * 24);
      if (isBefore(scheduledAt, firstDoseAt) && !isEqual(scheduledAt, firstDoseAt)) {
        continue;
      }
      if (scheduledAt > endsAt) {
        return doses.sort((a, b) => a.scheduledAt.getTime() - b.scheduledAt.getTime());
      }
      pushDose(scheduledAt);
      addedInDay += 1;
    }
    if (addedInDay === 0 && dayOffset > 0) {
      break;
    }
    dayOffset += 1;
    if (dayOffset > 366 || doses.length > 500) break;
  }

  return doses.sort((a, b) => a.scheduledAt.getTime() - b.scheduledAt.getTime());
}

File: backend/app/domain/doses.py
from __future__ import annotations

from datetime import datetime, timedelta

from app.domain.models import ParsedTreatment, PlannedDose


def build_dose_plan(parsed: ParsedTreatment) -> list[PlannedDose]:
    first = parsed.first_dose_at
    ends = parsed.ends_at
    if ends <= first:
        raise ValueError("A data de fim precisa ser depois da primeira dose.")

    remind_before = timedelta(minutes=parsed.remind_before_minutes)
    doses: list[PlannedDose] = []

    def push(scheduled_at: datetime) -> None:
        if scheduled_at > ends:
            return
        if any(d.scheduled_at == scheduled_at for d in doses):
            return
        doses.append(
            PlannedDose(
                scheduled_at=scheduled_at,
                remind_at=scheduled_at - remind_before,
            )
        )

    if parsed.frequency_every_hours:
        cursor = first
        while cursor <= ends and len(doses) < 500:
            push(cursor)
            cursor = cursor + timedelta(hours=parsed.frequency_every_hours)
        return sorted(doses, key=lambda d: d.scheduled_at)

    anchors = parsed.anchor_times or [first.strftime("%H:%M")]
    times_per_day = parsed.times_per_day or len(anchors)
    daily = anchors[: max(times_per_day, 1)]

    day_offset = 0
    while len(doses) < 500 and day_offset <= 366:
        added = 0
        for hhmm in daily:
            hour, minute = map(int, hhmm.split(":"))
            scheduled = first.replace(hour=hour, minute=minute, second=0, microsecond=0) + timedelta(
                days=day_offset
            )
            if scheduled < first:
                continue
            if scheduled > ends:
                return sorted(doses, key=lambda d: d.scheduled_at)
            push(scheduled)
            added += 1
        if added == 0 and day_offset > 0:
            break
        day_offset += 1

    return sorted(doses, key=lambda d: d.scheduled_at)


def estimate_dose_count(parsed: ParsedTreatment) -> int:
    return len(build_dose_plan(parsed))
